File: api/src/main/java/gg/mineral/api/math/MathUtil.java
package gg.mineral.api.math;

public interface MathUtil {
    default int toFixedPointInt(double value) {
        return (int) (value * 32);
    }

    default byte toFixedPointByte(double value) {
        return (byte) (value * 32);
    }

    default int floor(double num) {
        final int floor = (int) num;
        return floor == num ? floor : floor - (int) (Double.doubleToRawLongBits(num) >>> 63);
    }

    default int ceil(final double num) {
        final int floor = (int) num;
        return floor == num ? floor : floor + (int) (~Double.doubleToRawLongBits(num) >>> 63);
    }

    default int round(double num) {
        return floor(num + 0.5d);
    }

    default double square(double num) {
        return num * num;
    }

    default double sin(double num) {
        return Math.sin(num);
    }

    default double cos(double num) {
        return Math.cos(num);
    }

    default float toRadians(float angle) {
        return (float) Math.toRadians(angle);
    }

    default int toInt(Object object) {
        if (object instanceof Number number)
            return number.intValue();

        if (object == null)
            return 0;

        try {
            return Integer.valueOf(object.toString());
        } catch (NumberFormatException e) {
            e.printStackTrace();
        }
        return 0;
    }

    default float toFloat(Object object) {
        if (object instanceof Number number)
            return number.floatValue();

        if (object == null)
            return 0;

        try {
            return Float.valueOf(object.toString());
        } catch (NumberFormatException e) {
            e.printStackTrace();
        }
        return 0;
    }

    default double toDouble(Object object) {
        if (object instanceof Number number)
            return number.doubleValue();

        if (object == null)
            return 0;

        try {
            return Double.valueOf(object.toString());
        } catch (NumberFormatException e) {
            e.printStackTrace();
        }
        return 0;
    }

    default long toLong(Object object) {
        if (object instanceof Number number)
            return number.longValue();

        if (object == null)
            return 0;

        try {
            return Long.valueOf(object.toString());
        } catch (NumberFormatException e) {
            e.printStackTrace();
        }
        return 0;
    }

    default short toShort(Object object) {
        if (object instanceof Number number)
            return number.shortValue();

        if (object == null)
            return 0;

        try {
            return Short.valueOf(object.toString());
        } catch (NumberFormatException e) {
            e.printStackTrace();
        }
        return 0;
    }

    default byte toByte(Object object) {
        if (object instanceof Number number)
            return number.byteValue();

        if (object == null)
            return 0;
        try {
            return Byte.valueOf(object.toString());
        } catch (NumberFormatException e) {
            e.printStackTrace();
        }
        return 0;
    }

    default byte angleToByte(float angle) {
        return (byte) (angle * 256 / 360);
    }

    default short toVelocityUnits(double value) {
        return (short) (value * 8000D);
    }

    default double fromVelocityUnits(short value) {
        return value / 8000D;
    }

    default int toSoundUnits(double value) {
        return (int) (value * 8);
    }

    default double fromSoundUnits(int value) {
        return value / 8D;
    }

    default short toPitchUnits(double value) {
        return (short) (value * 63);
    }

    default double fromPitchUnits(short value) {
        return value / 63D;
    }

    default int unsigned(short s) {
        return s & 0xFFFF;
    }
}

File: api/src/main/java/gg/mineral/api/math/MathUtil.kt
package gg.mineral.api.math

interface MathUtil {
    fun toFixedPointInt(value: Double): Int {
        return (value * 32).toInt()
    }

    fun toFixedPointByte(value: Double): Byte {
        return (value * 32).toInt().toByte()
    }

    fun floor(num: Double): Int {
        val floor = num.toInt()
        return if (floor.toDouble() == num) floor else floor - (java.lang.Double.doubleToRawLongBits(num) ushr 63).toInt()
    }

    fun ceil(num: Double): Int {
        val floor = num.toInt()
        return if (floor.toDouble() == num) floor else floor + (java.lang.Double.doubleToRawLongBits(num)
            .inv() ushr 63).toInt()
    }

    fun round(num: Double): Int {
        return floor(num + 0.5)
    }

    fun square(num: Double): Double {
        return num * num
    }

    fun sin(num: Double): Double {
        return kotlin.math.sin(num)
    }

    fun cos(num: Double): Double {
        return kotlin.math.cos(num)
    }

    fun toRadians(angle: Float): Float {
        return Math.toRadians(angle.toDouble()).toFloat()
    }

    fun toInt(`object`: Any?): Int {
        if (`object` is Number) return `object`.toInt()

        if (`object` == null) return 0

        try {
            return `object`.toString().toInt()
        } catch (e: NumberFormatException) {
            e.printStackTrace()
        }
        return 0
    }

    fun toFloat(`object`: Any?): Float {
        if (`object` is Number) return `object`.toFloat()

        if (`object` == null) return 0f

        try {
            return `object`.toString().toFloat()
        } catch (e: NumberFormatException) {
            e.printStackTrace()
        }
        return 0f
    }

    fun toDouble(`object`: Any?): Double {
        if (`object` is Number) return `object`.toDouble()

        if (`object` == null) return 0.0

        try {
            return `object`.toString().toDouble()
        } catch (e: NumberFormatException) {
            e.printStackTrace()
        }
        return 0.0
    }

    fun toLong(`object`: Any?): Long {
        if (`object` is Number) return `object`.toLong()

        if (`object` == null) return 0

        try {
            return `object`.toString().toLong()
        } catch (e: NumberFormatException) {
            e.printStackTrace()
        }
        return 0
    }

    fun toShort(`object`: Any?): Short {
        if (`object` is Number) return `object`.toShort()

        if (`object` == null) return 0

        try {
            return `object`.toString().toShort()
        } catch (e: NumberFormatException) {
            e.printStackTrace()
        }
        return 0
    }

    fun toByte(`object`: Any?): Byte {
        if (`object` is Number) return `object`.toByte()

        if (`object` == null) return 0
        try {
            return `object`.toString().toByte()
        } catch (e: NumberFormatException) {
            e.printStackTrace()
        }
        return 0
    }

    fun angleToByte(angle: Float): Byte {
        return (angle * 256 / 360).toInt().toByte()
    }

    fun toVelocityUnits(value: Double): Short {
        return (value * 8000.0).toInt().toShort()
    }

    fun fromVelocityUnits(value: Short): Double {
        return value / 8000.0
    }

    fun toSoundUnits(value: Double): Int {
        return (value * 8).toInt()
    }

    fun fromSoundUnits(value: Int): Double {
        return value / 8.0
    }

    fun toPitchUnits(value: Double): Short {
        return (value * 63).toInt().toShort()
    }

    fun fromPitchUnits(value: Short): Double {
        return value / 63.0
    }

    fun unsigned(s: Short): Int {
        return s.toInt() and 0xFFFF
    }
}
